app/src/main/java/org/schabi/newpipe/player/datasource/RefreshableHlsHttpDataSource.java

```diff
@@ -0,0 +1,248 @@
+package org.schabi.newpipe.player.datasource;
+
+import static org.schabi.newpipe.MainActivity.DEBUG;
+
+import android.net.Uri;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.google.android.exoplayer2.source.hls.playlist.HlsMediaPlaylist;
+import com.google.android.exoplayer2.source.hls.playlist.HlsPlaylistParser;
+import com.google.android.exoplayer2.upstream.DataSourceInputStream;
+import com.google.android.exoplayer2.upstream.DataSpec;
+import com.google.android.exoplayer2.upstream.HttpDataSource;
+
+import org.schabi.newpipe.extractor.exceptions.ExtractionException;
+import org.schabi.newpipe.extractor.stream.RefreshableStream;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class RefreshableHlsHttpDataSource extends LoggingHttpDataSource {

```

Comment:
Add some JDoc about why a refreshable data source is needed and how this is achieved. We know, but people who are new to the project might not.